h2/src/main/org/h2/result/LocalResult.java

```diff
@@ -368,7 +369,7 @@ public void done() {
             }
             if (sort != null) {
                 if (offset > 0 || limit > 0) {
-                    sort.sort(rows, offset, limit < 0 ? rows.size() : limit);
+                    sort.sort(rows, offset, limit < 0 || withTies ? rows.size() : limit);
```

Comment:
this check looks like it should be part of the if statement on the previous line

Also that limit < 0 in combination with the limit > 0 on the previous line looks very dodgy

Reply:
Possible situations:
1. `OFFSET > 0` with any `LIMIT`. Partial sorting is used.
2. `LIMIT > 0` with any `OFFSET`. Partial sorting is used.
3. If `LIMIT = 0` sorting is not useful, but this situation is abnormal and was not checked here, full sorting is used.
4. All other cases should trigger full sorting.

If partial sorting is used due to (1) or (2) `OFFSET` is always known (may also be 0). Situation with `LIMIT` is more complicated. If `LIMIT < 0` limit is not in use so all remaining rows should be sorted, the same for `WITH TIES` condition because it will read extra rows. In other cases `LIMIT` is used as is.

So from my point of view logic is fine here, but `LIMIT = 0` can be checked separately with quick return.

Reply:
ok

Reply:
I slightly reorganized conditions to make logic more readable and also removed sorting in (3) because all rows (if any) will be removed in `applyOffsetAndLimit()` in this situation.